File: Algorithms/C++/Medium/71_Simplify_Path.cpp
class Solution {
public:
    string simplifyPath(string path) {
        if (path.empty()) {
            return "";
        }

        vector<string> pathComponents = split(path, '/');

        stack<string> reversePathStack;
        for (int i = 0; i < pathComponents.size(); i++) {
            string component = pathComponents[i];
            if (component == ".") {
                ;
            } else if (component == "..") {
                if (!reversePathStack.empty()) {
                    reversePathStack.pop();
                }
            } else {
                reversePathStack.push(component);
            }
        }

        if (reversePathStack.empty()) {
            return "/";
        }

        stack<string> pathStack;
        while (!reversePathStack.empty()) {
            pathStack.push(reversePathStack.top());
            reversePathStack.pop();
        }

        string simplifiedPath = "";
        while (!pathStack.empty()) {
            simplifiedPath += "/" + pathStack.top();
            pathStack.pop();
        }

        return simplifiedPath;
    }

private:
    vector<string> split(const string &str, char delimiter) {
        vector<string> strs;

        stringstream ss(str);
        string item;

        while (getline(ss, item, delimiter)) {
            if (!item.empty()) {
                strs.push_back(item);
            }
        }

        return strs;
    }
};

File: Algorithms/JavaScript/Medium/71_Simplify_Path.js
/**
 * @param {string} path
 * @return {string}
 */
var simplifyPath = function(path) {
    if (path === "") {
        return "";
    }

    var pathComponents = path.split("/").filter(function(str) {
        return str !== "";
    });

    if (pathComponents.length === 0) {
        return "/";
    }

    var componentsStack = [], i = 0;
    for (i = 0; i < pathComponents.length; i++) {
        var component = pathComponents[i]
        if (component === ".") {

        } else if (component === "..") {
            if (componentsStack.length > 0) {
                componentsStack.pop();
            }
        } else {
            componentsStack.push(component);
        }
    }

    var simplifiedPath = "";
    if (componentsStack.length === 0) {
        simplifiedPath = "/";
    } else {
        for (i = 0; i < componentsStack.length; i++) {
            simplifiedPath += "/" + componentsStack[i];
        }
    }

    return simplifiedPath;
};
